RecoParticleFlow/PFProducer/plugins/MLPFProducerSonic.cc

```diff
@@ -0,0 +1,199 @@
+#include "HeterogeneousCore/SonicCore/interface/SonicEDProducer.h"
+#include "HeterogeneousCore/SonicTriton/interface/TritonClient.h"
+#include "RecoParticleFlow/PFProducer/interface/MLPFModel.h"
+
+#include "TLorentzVector.h"
+
+// git clone https://github.com/jpata/mlpf-models.git
+// mv mlpf-models/triton/models ./
+// singularity exec --nv docker://nvcr.io/nvidia/tritonserver:20.10-py3 tritonserver --model-repository models --log-verbose=1 --log-error=1 --log-info=1
+//
+// process.mlpfproducergpu = cms.EDProducer("MLPFProducerSonicTriton",
+//     Client = cms.PSet(
+//         mode = cms.string("Sync"),
+//         batchSize = cms.untracked.uint32(1),
+//         address = cms.untracked.string("127.0.0.1"),
+//         port = cms.untracked.uint32(8001),
+//         timeout = cms.untracked.uint32(60),
+//         modelName = cms.string("mlpf"),
+//         modelVersion = cms.string("1"),
+//         verbose = cms.untracked.bool(False),
+//         allowedTries = cms.untracked.uint32(0),
+//     )
+// )
+
+class MLPFProducerSonicTriton : public SonicEDProducer<TritonClient> {
+public:
+  //needed because base class has dependent scope
+  explicit MLPFProducerSonicTriton(edm::ParameterSet const& cfg)
+      : SonicEDProducer<TritonClient>(cfg),
+        pfCandidatesToken_{this->template produces<reco::PFCandidateCollection>()},
+        inputTagBlocks_(this->template consumes<reco::PFBlockCollection>(cfg.getParameter<edm::InputTag>("src"))) {
+    //for debugging
+    this->setDebugName("MLPFProducerSonic");
+  }
+
+  void acquire(edm::Event const& event, edm::EventSetup const& iSetup, Input& iInput) override {
+    edm::LogInfo(client_.debugName()) << "acquire";
+
+    using namespace reco::mlpf;
+
+    auto blocks_handle = event.getHandle(inputTagBlocks_);
+    const auto& blocks = *blocks_handle;
+    std::vector<reco::PFCandidate> pOutputCandidateCollection;
+
+    std::vector<const reco::PFBlockElement*> all_elements;
+    for (const auto& block : blocks) {
+      const auto& elems = block.elements();
+      for (const auto& elem : elems) {
+        if (all_elements.size() < NUM_MAX_ELEMENTS_BATCH) {
+          all_elements.push_back(&elem);
+        } else {
+          //model needs to be created with a bigger LSH codebook size
+          edm::LogError("MLPFProducer") << "too many input PFElements for predefined model size: " << elems.size();
+          break;
+        }
+      }
+    }
+    const long long int num_elements_total = all_elements.size();
+
+    //tensor size must be a multiple of the bin size and larger than the number of elements
+    const long long int tensor_size = LSH_BIN_SIZE * (num_elements_total / LSH_BIN_SIZE + 1);
+    assert(tensor_size <= NUM_MAX_ELEMENTS_BATCH);
+
+    auto& input1 = iInput.at("x");
+
+    //we ignore Sonic/Triton batching, as it doesn't create a dim-3 input for batch size 1.
+    //instead, we specify the batch dim as a model dim.
+    input1.setShape(0, 1);
```

Comment:
https://github.com/jpata/mlpf-models/blob/main/triton/models/mlpf/config.pbtxt currently just has 2 dimensions for input.

In general, only dimensions that are `-1` in the config.pbtxt should be set by the user in `acquire()`.

Reply:
I forgot to push the provisionary mlpf-models repo to the latest version. Per private discussions, we need to specify the batch dimension explicitly as a model dimension, as triton tries to run the model without the batch dim otherwise, which is not supported by the MLPF model.